drv/lpc55-swd/src/main.rs

```diff
@@ -1748,195 +1725,151 @@ impl ServerImpl {
+            // If the write fails, then attempt to undo it
+            ringbuf_entry!(Trace::DhcsrWriteError);
+            self.disable_halting_debug();
+            return Err(());
+        }
 
-            self.sp_reset_leave();
-            need_undo &= !Undo::RESET;
-
-            // 500ms max wait allows for testing using manual reset button.
-            // Typical wait looks to be 5ms.
-            self.wait_for_sp_halt(WAIT_FOR_HALT_MS).map_err(|_| ())?;
-
-            // Check that RESET was caught
-            if let Ok(dfsr) = self.dp_read_bitflags::<Dfsr>() {
-                if !dfsr.is_vcatch() {
-                    ringbuf_entry!(Trace::Dfsr(dfsr));
-                    ringbuf_entry!(Trace::VcCoreResetNotCaught);
-                    return Err(());
-                }
-            } else {
-                ringbuf_entry!(Trace::DfsrReadError);
-            }
+        // Enable vector catch
+        if self
+            .dp_write_bitflags::<Demcr>(Demcr::VC_CORERESET)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::DemcrWriteError);
+            self.disable_halting_debug();
+            self.disable_vector_catch();
+            return Err(());
+        }
+        self.sp_reset_leave();
 
-            // We don't want to catch the next reset.
-            if self
-                .dp_write_bitflags::<Demcr>(Demcr::from_bits_retain(0))
-                .is_err()
-            {
-                ringbuf_entry!(Trace::DemcrWriteError);
-                return Err(());
-            }
+        // 500ms max wait allows for testing using manual reset button.
+        // Typical wait looks to be 5ms.
+        let halted = self.wait_for_sp_halt(WAIT_FOR_HALT_MS);
 
-            // need_undo was set appropriately
-            if let Ok(demcr) = self.dp_read_bitflags::<Demcr>() {
-                if demcr & Demcr::VC_CORERESET != Demcr::VC_CORERESET {
-                    ringbuf_entry!(Trace::VcCoreReset(false));
-                    need_undo &= !Undo::VC_CORERESET;
+        // We always disable vector catch, because we want the next reset to
+        // proceed as usual.
+        self.disable_vector_catch();
+        let out = match halted {
+            Ok(()) => {
+                // Check that RESET was caught; if we halted for some other
+                // reason (or can't read DFSR), then that's a bad sign.
+                if let Ok(dfsr) = self.dp_read_bitflags::<Dfsr>() {
+                    if dfsr.is_vcatch() {
+                        Ok(())
+                    } else {
+                        ringbuf_entry!(Trace::Dfsr(dfsr));
+                        ringbuf_entry!(Trace::VcCoreResetNotCaught);
+                        Err(())
+                    }
                 } else {
-                    ringbuf_entry!(Trace::VcCoreReset(true));
-                    return Err(());
+                    ringbuf_entry!(Trace::DfsrReadError);
+                    Err(())
```

Comment:
take it or leave it: this could also be written as:
```rust
                // Check that RESET was caught; if we halted for some other
                // reason (or can't read DFSR), then that's a bad sign.
                match self.dp_read_bitflags::<Dfsr>() {
                    Ok(dsfr) if dfsr.is_vcatch() => Ok(()),
                    Ok(dsfr) => {
                        ringbuf_entry!(Trace::Dfsr(dfsr));
                        ringbuf_entry!(Trace::VcCoreResetNotCaught);
                        Err(())
                    },
                    Err(_) => {
                        ringbuf_entry!(Trace::DfsrReadError);
                        Err(())
                    }
                }
```
though it's up to you whether you think that's any nicer

Reply:
It's exactly the same LoC, but I slightly prefer it, thanks!